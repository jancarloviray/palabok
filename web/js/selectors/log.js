/* @flow */

/*::
  import type { State } from './state'
  import type { Labels } from './label'
  import type { TimerType } from './timer'

  export type Logs = {
    [key: string]: Log | null
  }

  export type Log = {
    id: string,
    timerType: TimerType,
    startedAt: Date,
    endedAt: Date,
    duration: number,
    labelId: string,
    isComplete: boolean
  }

  export type FullLog = {
    id: string,
    timerType: TimerType,
    startedAt: Date,
    endedAt: Date,
    duration: number,
    labelId: string,
    isComplete: boolean,
    labelText: string
  }

  export type LogsByDate = {
    [key: string]: Array<Log>
  }
*/

import filter from 'object-loops/filter'
import get from '101/pluck'
import groupBy from '101/group-by'
import values from 'object-loops/values'
import { createSelector } from 'reselect'
import { full as fullLabel } from './label'

/*
 * Returns midnight
 */

export const midnight /*: (state: State) => Date */ = createSelector(
  state => get(state, 'time.now'),
  now => truncateDate(now))

/*
 * Recent work logs
 */

export const recents /*: (state: State) => Logs */ = createSelector(
  state => state.log,
  state => midnight(state),
  (log, time) => {
    log = onlyWork(log)
    log = filter(log, i => i.startedAt > time)
    return log
  })

/*
 * Full
 */

export const full /*: ([Log, Labels]) => FullLog */ = createSelector(
  ([log, _]) => log,
  ([log, labels]) => {
    return fullLabel(log.labelId && labels[log.labelId])
  },
  (log, label) => {
    return {
      ...log,
      labelText: log.timerType === 'work' ? label.name : 'Break'
    }
  })

/*
 * By date
 */

export const byDate /*: (state: State) => LogsByDate */ = createSelector(
  state => state.log || {},
  log => {
    return groupBy(values(log).filter(Boolean), item =>
      truncateDate(item.startedAt).toISOString())
  })

/*
 * Only work
 */

export const onlyWork /*: (logs: Logs) => Logs */ = createSelector(
  (log /*: Log */) => log || {},
  (log /*: Log */) => {
    return filter(log, item => item && item.timerType === 'work')
  })

/**
 * Truncates a date to midnight.
 */

export function truncateDate (date /*: Date */) {
  date = new Date(date)
  date.setMilliseconds(0)
  date.setSeconds(0)
  date.setMinutes(0)
  date.setHours(0)
  return date
}

/*
 * Export
 */

export default {
  recents,
  byDate,
  onlyWork
}
