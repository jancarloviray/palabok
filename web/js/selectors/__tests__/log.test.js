/* eslint-env jest */
import { byDate, recents } from '../log'

describe('byDate', () => {
  test('works on empty logs', () => {
    const result = byDate({
      log: {}
    })

    expect(result).toMatchSnapshot()
  })

  test('works on undefined logs', () => {
    const result = byDate({})
    expect(result).toMatchSnapshot()
  })

  test('groups by date', () => {
    const src = {
      log: {
        a: {
          id: 'a',
          startedAt: new Date('2010-04-20T08:00:00Z')
        },
        b: {
          id: 'b',
          startedAt: new Date('2010-04-21T08:00:00Z')
        },
        c: {
          id: 'c',
          startedAt: new Date('2010-04-21T09:00:00Z')
        }
      }
    }

    const result = byDate(src)
    const keys = Object.keys(result)
    expect(keys.length).toEqual(2)
    expect(result[keys[0]].length).toEqual(1)
    expect(result[keys[1]].length).toEqual(2)
  })
})

describe('recents', () => {
  test('works on empty logs', () => {
    const result = recents({
      log: {}
    })

    expect(result).toMatchSnapshot()
  })

  test('returns just todays items', () => {
    const result = recents({
      time: {
        now: new Date('2010-04-21T08:00:00Z')
      },
      log: {
        a: {
          id: 'a',
          timerType: 'work',
          startedAt: new Date('2010-04-20T08:00:00Z')
        },
        b: {
          id: 'b',
          timerType: 'work',
          startedAt: new Date('2010-04-21T08:00:00Z')
        },
        c: {
          id: 'c',
          timerType: 'work',
          startedAt: new Date('2010-04-21T09:00:00Z')
        }
      }
    })

    expect(result).toMatchSnapshot()
  })
})
