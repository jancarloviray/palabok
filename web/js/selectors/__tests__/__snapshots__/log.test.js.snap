// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`byDate works on empty logs 1`] = `Object {}`;

exports[`byDate works on undefined logs 1`] = `Object {}`;

exports[`recents returns just todays items 1`] = `
Object {
  "b": Object {
    "id": "b",
    "startedAt": 2010-04-21T08:00:00.000Z,
    "timerType": "work",
  },
  "c": Object {
    "id": "c",
    "startedAt": 2010-04-21T09:00:00.000Z,
    "timerType": "work",
  },
}
`;

exports[`recents works on empty logs 1`] = `Object {}`;
